### PR TITLE
fix: workflow status update and resource step can not upgrade 

### PR DESCRIPTION
**Pull request by alexcodelf**
**Problem:**

- can not upgrade resource in workflow
- a pending step will not change status when cancelling workflow.
- workflow execution status will not changed to error when sub step failed.

**Solution:**

- add resource definition matching rules for resource step.
- omitted step will update step status
- fail workflow fast when step failed

**Related Issue:**
#1955 #1956 #1957

### CODE REVIEW

File: pkg/workflow/step/service_step.go [hunk 1]
```diff
@@ -199,3 +201,35 @@ func (s *ServiceStepManager) GenerateTemplates(
...
+		Where(resource.Name(name)).
+		WithResourceDefinitionMatchingRule().
+		Only(ctx)
+	if err != nil && !model.IsNotFound(err) {
+		return nil, err
+	}
+
+	if res != nil && res.Edges.ResourceDefinitionMatchingRule != nil {
+		stepAttrs["resourceDefinitionMatchingRule"] = map[string]any{
+			"id": res.Edges.ResourceDefinitionMatchingRule.ID.String(),
```
Comment (gitlawr) [inline pkg/workflow/step/service_step.go:230]: How do we use this?

Comment (alexcodelf) [reply pkg/workflow/step/service_step.go:230]: Resource validation need `ResourceDefinitionMatchingRuleID` or `templateID` to generate computed attributes, will need add this for resource upgrade API.

File: pkg/apis/resource/extension_view.go [hunk 2]
```diff
@@ -530,15 +533,10 @@ func (r *CollectionRouteUpgradeRequest) Validate() error {
...
-		} else {
-			en := updateInputsItemToResource(entity.Type, input, r.Project, r.Environment)
-
-			input.ComputedAttributes, err = pkgresource.GenComputedAttributes(r.Context, r.Client, en)
-			if err != nil {
-				return err
-			}
 		}
 
+		en := updateInputsItemToResource(entity.Type, input, r.Project, r.Environment)
```
Comment (gitlawr) [inline pkg/apis/resource/extension_view.go:538]: We don't call this if not ReuseAttributes before. Is this expected change?

Comment (alexcodelf) [reply pkg/apis/resource/extension_view.go:538]: This variable Is only used in computed resources, the func will new `*model.Resource`, the input data will not be effected.

**Review by gitlawr — APPROVED** [review]
LGTM

File: pkg/apis/resource/extension_view.go [hunk 3]
```diff
@@ -176,10 +168,21 @@ func (r *RouteUpgradeRequest) Validate() error {
...
 
+	if !r.ReuseAttributes {
+		computedAttr, err := pkgresource.GenComputedAttributes(r.Context, r.Client, en)
+		if err != nil {
+			return err
+		}
+
+		r.ComputedAttributes = computedAttr
+	}
+
```
Comment (thxCode) [inline pkg/apis/resource/extension_view.go:185]: we should make `if-else` look like a whole logic, please put the above `if` block here too.

```
if r.ReuseAttributes {
		r.Attributes = entity.Attributes
		r.ComputedAttributes = entity.ComputedAttributes
		if r.Labels == nil {
			r.Labels = entity.Labels
		}
} else {
                computedAttr, err := pkgresource.GenComputedAttributes(r.Context, r.Client, en)
		if err != nil {
			return err
		}

		r.ComputedAttributes = computedAttr
}
```

**Review by thxCode — APPROVED** [review]
LGTM

**Pull request merged**